stream setting: Switch webapp to subscribe/unsubscribe from streams by ID.

Riken-Shah opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Modified subscribe/unsubscribe API to select streams from either `id` or `name` 
using a common attribute `stream`.

Also changing frontend to subscribe/unsubscribe streams by id.
Fixes: #10744


**Testing plan:** <!-- How have you tested? -->
Modified and add a test to get 100% test coverage.



**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

abhijeetbodas2001 commented on zerver/lib/streams.py:
```diff
@@ -131,26 +132,34 @@ def create_streams_if_needed(
+                existing_streams.append(stream)
         else:
-            existing_streams.append(stream)
+            # Raising error if unknown stream id is passed.
+            try:
+                stream = get_stream_by_id_in_realm(stream_dict["stream"], realm)
+            except Stream.DoesNotExist:
+                raise JsonableError(_("No stream with id '{}' found.").format(stream_dict["stream"]))
```
Rather than raising the error here, I think it would be better to enforce that `autocreate=True` is never passed to `list_to_streams` when using `stream_id` as identifier. (Maybe raise an error in `list_to_streams` itself if this is not followed?)

Riken-Shah replied:
Yes @abhijeetbodas2001,  I agree with you it will be better this way.

abhijeetbodas2001 commented on zerver/lib/streams.py:
```diff
@@ -511,17 +520,28 @@ def list_to_streams(streams_raw: Iterable[StreamDict],
     existing_streams: List[Stream] = []
     missing_stream_dicts: List[StreamDict] = []
-    existing_stream_map = bulk_get_streams_by_names(user_profile.realm, stream_set)
+    existing_stream_map = bulk_get_streams_by_names(user_profile.realm, cast(Set[str], stream_set)) if is_stream_type_name else bulk_get_streams_by_ids(user_profile.realm, cast(Set[int], stream_set))
+    # Casting it to Dict[Union[int, str], Any] beacuse inline if/else
+    # (foo = x if y else z) returns object.
+    existing_stream_map = cast(Dict[Union[int, str], Any], existing_stream_map)
```
Would it be better to just use the normal `if else` syntax here? (Since the line length is also quiet long here)

abhijeetbodas2001 commented on zerver/lib/streams.py:
```diff
@@ -511,17 +520,28 @@ def list_to_streams(streams_raw: Iterable[StreamDict],
-        check_stream_name(stream_name)
+    else:
+        for stream_id in stream_set:
+            # Stream id should already be parsed in int by the caller,
```
All `dict`s in `streams_raw` must either have the name or the id as `stream` paramter correct? In that case, I think we should specify that clearly in the docstring of `list_to_streams`

abhijeetbodas2001 replied:
Also, FWIW how about using a more descriptive name for the `StreamDict` paramter than just `stream`?

Riken-Shah replied:
What do you think will be a better name for it?

Riken-Shah replied:
My suggestion: `stream`, `stream_identifer`.

abhijeetbodas2001 replied:
Hmm, I'm not really sure what a good name for it could be. Anything involving "identifier" could be confused with `id`.
I pointed this out because `stream` is also used to indicate the name in other parts of the codebase.
Maybe it would be best to somehow have both  `id` and `name` as keys in `StreamDict`?

Riken-Shah replied:
Actually, Tim suggested that there should be only one parameter to represent both the stream name and stream id.
(https://github.com/zulip/zulip/issues/10744#issuecomment-434875243).
IDK If it helps but I have added a comment in `StreamDict` that the `stream` attribute is used to represent both.

timabbott replied:
I think if we're passing a dict, we should use different parameter names; I had been thinking about a "list" context.

Riken-Shah replied:
@timabbott  Can you please explain a bit about what `list context` means?

timabbott replied:
I'd been imagining a list of things that are either a string or an integer, not a dictionary.

abhijeetbodas2001 commented:
@Riken-Shah Sorry for the delay, I missed this. You can mention/comment after updating a PR so that GitHub sends a notification. Some questions/comments-

1. Do we want to rename other existing functions to say that they accept stream names? For example, `models.py/get_realm_stream` and `models.py/get_stream`?
2. I think a commit structure like this would be cleaner and make this easier to review-
   1. All the renaming changes (adding `by_name`)
   2. All the code related to getting streams by IDs (the cache key function and the `bulk_get_by_id`
   3. The backend changes related to the API and documentation
   4. Client-side changes to use the API with IDs

    The idea is, all changes related to adding/changing one functionality should be put together into one commit, without it breaking anything. What do you think?

The code changes look good to me 🙂

Riken-Shah committed:
refactor: Rename some stream function to say they accept the name.

This commit renames some functions which have a
similar function with `stream_id`(for eg:
`bulk_get_streams` and `bulk_get_streams_by_ids`) for
more clarity.

Riken-Shah commented:
>     1. Do we want to rename other existing functions to say that they accept stream names? For example, `models.py/get_realm_stream` and `models.py/get_stream`?
I don't think so it's necessary as long as there are two similar functions that use stream name and stream id respectively.
Eg: `bulk_get_streams` and `bulk_get_streams_by_ids`.
I have added a [commit](https://github.com/zulip/zulip/pull/16899/commits/ff663b68ed1aff8bb069e64f86f3fc48cc21190f) for this.


>     2. I think a commit structure like this would be cleaner and make this easier to review-
>        
>        1. All the renaming changes (adding `by_name`)
>        2. All the code related to getting streams by IDs (the cache key function and the `bulk_get_by_id`
>        3. The backend changes related to the API and documentation
>        4. Client-side changes to use the API with IDs

I have updated the commit structure as you suggested but 3 and 4 cannot be separated because we are replacing (stream)`name` with `stream` in the APIs, which will lead to puppeteer tests failing if we change the frontend part in a separate commit.

Riken-Shah commented:
@abhijeetbodas2001 Updated the PR!

abhijeetbodas2001 commented:
> 3 and 4 cannot be separated because we are replacing (stream)name with stream in the APIs, which will lead to puppeteer tests failing if we change the frontend part in a separate commit.

Makes sense. Something like this instead, if I'm not wrong, won't break tests right?
3. Rename `name`->`stream` in API as well as JS code
4. Server changes to support IDs
5. Frontend changes to send IDs in the `stream` paramter instead of sending stream names

But maybe this is too tangled? I'm not sure. Let's wait for Tim to comment here.

I think the 2nd commit summary should be more specific. `refactor: Rename some function for better interpretation.` doesn't really say you're dealing with streams-related code. I think something along the lines of `refactor: Rename stream API functions to say they accept names.` would be better. And then you could explain why we're doing this in the commit body.

Riken-Shah commented:
@zulipbot add "needs review"

timabbott commented:
@Riken-Shah can you rebase this, now that 4.0 is released?

Riken-Shah committed:
cache: Add function to generate cache key for stream id + realm id.

This is a prep commit that will help us allow streams to
subscribe/unsubscribe via id.

Riken-Shah committed:
stream: Add function to bulk fetch streams via ids (cache enable).

This is a prep commit that will help us allow streams to
subscribe/unsubscribe via id.

Riken-Shah commented:
@timabbott  I have rebased my PR on top of my master. 
Also improved the commit messages.

I'll update the PR again after my [doubt](https://github.com/zulip/zulip/pull/16899#discussion_r633054184) is resolved.

Riken-Shah committed:
streams: Allow streams to subscribe/unsubscribe via stream name or id.

This commit adds frontend and backend changes to
allow streams to subscribe/unsubscribe via stream id.

Fixes: #10744
